### PR TITLE
Discord support

### PR DESCRIPTION
**Pull request by askovpen**
ready to use

### CODE REVIEW

**Comment by DBa2016** [comment]
What does it do? Can you document it in the appropriate file in the docs directory?

**Comment by askovpen** [comment]
While - for testing - not merge :-)

**Comment by sohje** [comment]
https://discordapp.com/ ?

**Comment by DBa2016** [comment]
@sohje - I know roughly what discord is, but what does it do here? Notification only? Controlling the bot?...

**Comment by askovpen** [comment]
@DBa2016 notify and controlling

File: pokemongo_bot/event_handlers/discord_handler.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        except ValueError as exception:
+            self.bot.logger.info('[x] Error while opening inventory file for read: %s' % exception)
+            json_inventory = []
+        except:
+            raise FileIOException("Unexpected error reading from {}".format(web_inventory))
+        return next((x["inventory_item_data"]["player_stats"]
+                     for x in json_inventory
+                     if x.get("inventory_item_data", {}).get("player_stats", {})),
+                    None)
+
```
Comment (Gobberwart) [inline pokemongo_bot/event_handlers/discord_handler.py]: There is no need to reload inventory from json. Just access the appropriate Inventory (inventory.player_stats from memory)

**Comment by Gobberwart** [comment]
You will end up duplicating a lot of code here to give discord the same functionality as telegram.

This is why I have suggested that the code to generate events and respond to commands be separated from the actual communication method. 

So we could have discordbot and telegrambot that just handle the communication aspect of things, while another class passes information to/from these bots.

**Comment by Gobberwart** [comment]
Love the concept though. We should not have to be forced to use Telegram just to get this functionality.

Would like to see handlers for Telegram, Discord, Email, Other common messaging apps, all sharing the same basic alert generation/command response base code.

File: pokemongo_bot/event_handlers/discord_handler.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                    "*"+self.bot.config.username+"*",
+                    "_Level:_ "+str(stats["level"]),
+                    "_XP:_ "+str(stats["experience"])+"/"+str(stats["next_level_xp"]),
+                    "_Pokemons Captured:_ "+str(stats["pokemons_captured"])+" ("+str(catch_day)+" _last 24h_)",
+                    "_Poke Stop Visits:_ "+str(stats["poke_stop_visits"])+" ("+str(ps_day)+" _last 24h_)",
+                    "_KM Walked:_ "+str("%.2f" % stats["km_walked"])
+                )
+            self.sendMessage(to=chat_id, parse_mode='Markdown', text="\n".join(res))
+        else:
+            self.sendMessage(to=chat_id, parse_mode='Markdown', text="Stats not loaded yet\n")
```
Comment (Gobberwart) [inline pokemongo_bot/event_handlers/discord_handler.py]: This is what I mean about duplicate code. Telegram handler does the same thing. Why not separate out and let both Telegram and Discord (and potentially other handlers in future) share?

File: pokemongo_bot/event_handlers/discord_handler.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            return True
+        try:
+            oper = re.search(r'operator:([^ ]+)', params).group(1)
+            rule_cp = int(re.search(r'cp:([0-9]+)', params).group(1))
+            rule_iv = float(re.search(r'iv:([0-9.]+)', params).group(1))
+            rule_pkmn = re.search(r'pokemon:([^ ]+)', params).group(1)
+            return rule_pkmn == pokemon and (oper == "or" and (cp >= rule_cp or iv >= rule_iv) or cp >= rule_cp and iv >= rule_iv)
+        except:
+            return False
+
```
Comment (Gobberwart) [inline pokemongo_bot/event_handlers/discord_handler.py]: More duplicate code from Telegram

File: pokemongo_bot/event_handlers/discord_handler.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    def handle_event(self, event, sender, level, formatted_msg, data):
+        if self.dbot is None:
+            try:
+                self.bot.logger.info("Discord bot not running, trying to spin it up")
+                self.dbot = DiscordClass(self.bot, self.master, self.pokemons, self.config)
+                self.dbot.connect()
+                thread.start_new_thread(self.dbot.run)
+            except Exception as inst:
+                self.dbot = None
+                self.bot.logger.error("Unable to spin Telegram bot; master: {}, exception: {}".format(self.master, pprint.pformat(inst)))
```
Comment (sohje) [inline pokemongo_bot/event_handlers/discord_handler.py]: Telegram?

File: pokemongo_bot/event_handlers/discord_handler.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+
+DEBUG_ON = False
+
+class FileIOException(Exception):
+    pass
+
+class DiscordClass:
+
+    update_id = None
```
Comment (sohje) [inline pokemongo_bot/event_handlers/discord_handler.py]: any usage?

File: pokemongo_bot/event_handlers/discord_handler.py (hunk 2 above)
Comment (askovpen) [inline pokemongo_bot/event_handlers/discord_handler.py]: In future can rename "TelegramTask" config keys to some like "alerts" and add key: agents: ["telegram", "discord", "email"]

File: pokemongo_bot/event_handlers/discord_handler.py (hunk 1 above)
Comment (Gobberwart) [inline pokemongo_bot/event_handlers/discord_handler.py]: This is still unnecessarily convoluted. Why not just:

```
def _get_player_stats(self):
    return inventory.player().player_stats
```

File: pokemongo_bot/event_handlers/discord_handler.py [hunk 6]
```diff
@@ -78,7 +71,7 @@ def on_message(self,message):
 
     def run(self):
       self._dbot.forever_loop()
-
+      self = None
```
Comment (Gobberwart) [inline pokemongo_bot/event_handlers/discord_handler.py]: What?

**Comment by Gobberwart** [comment]
I have significant concerns about this, in that discord support duplicates a lot of the code in telegram_handler.py. Which means any time we want to change it, we have to change both.

OK to go ahead with this approach for now, but ultimately we should make discord and telegram handlers just look after communicating with discord/telegram, and leave the processing of commands and generation of events to a separate class.

**Comment by solderzzc** [comment]
Same as @Gobberwart , is there better way to organize code?

File: pokemongo_bot/event_handlers/discord_handler.py (hunk 1 above)
Comment (askovpen) [inline pokemongo_bot/event_handlers/discord_handler.py]: You view old code. View last

**Comment by DBa2016** [comment]
I think there are two options:
1) move the logic into a base class, from which telegram, discord, and whateverelse handlers can be derived. The handlers should then **only** include code specific to facilitate the communication between the handler and the messenger.
2) move the logic into "control handler" or similar class which processes events generated/emitted by the particular comunication class.

There might be other options which reduce code duplication, I just don't se them right now :)

Both approaches would even allow web interfaces to run over the same logic... As for email - I don't think this would be particularly useful, but it does not hurt either.

If we don't do this, every bug fix or enhancement we do in a control interface will need to be duplicated (including the logic!) to all other control interfaces.

If no objections are xpressed, I can start implementing option 2 - however, a code freeze for Telegram (and discord) code will be needed for this.

File: pokemongo_bot/event_handlers/discord_handler.py (hunk 6 above)
Comment (askovpen) [inline pokemongo_bot/event_handlers/discord_handler.py]: fixed

**Comment by askovpen** [comment]
approve PR and try make joining to your concept :-)

**Comment by solderzzc** [comment]
@DBa2016 Agreed, let's move on.

**Comment by solderzzc** [comment]
👍

**Comment by MZorzy** [comment]
./setup.sh -r

Traceback (most recent call last):
  File "pokecli.py", line 47, in <module>
    from pokemongo_bot import PokemonGoBot, TreeConfigBuilder
  File "/home/dev/Poke/PokemonGo-Bot/pokemongo_bot/__init__.py", line 22, in <module>
    import cell_workers
  File "/home/dev/Poke/PokemonGo-Bot/pokemongo_bot/cell_workers/__init__.py", line 26, in <module>
    from telegram_task import TelegramTask
  File "/home/dev/Poke/PokemonGo-Bot/pokemongo_bot/cell_workers/telegram_task.py", line 10, in <module>
    from pokemongo_bot.event_handlers import TelegramHandler
  File "/home/dev/Poke/PokemonGo-Bot/pokemongo_bot/event_handlers/__init__.py", line 5, in <module>
    from discord_handler import DiscordHandler
  File "/home/dev/Poke/PokemonGo-Bot/pokemongo_bot/event_handlers/discord_handler.py", line 7, in <module>
    import discord_simple
ImportError: No module named discord_simple

**Pull request merged**